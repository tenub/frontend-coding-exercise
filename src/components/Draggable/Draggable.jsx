import React from 'react';
import { DragSource } from 'react-dnd';

import { ItemTypes } from '../../ItemTypes';

class Draggable extends React.PureComponent {
  /**
   * Allow all items with a defined type to be draggable.
   */
  static type = ({ type }) => {
    const itemTypeKey = type.toUpperCase()
    return ItemTypes[itemTypeKey];
  }

  /**
   * Share necessary item properties so they can be used in determining
   * draggable and droppable states.
   */
  static spec = {
    beginDrag: (props, monitor, component) => {
      const { children, index, name, type } = props;
      return { children, index, name, type };
    },
  }

  static collect = (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  })

  render() {
    const { connectDragSource, isDragging, children } = this.props;
    const opacity = isDragging ? 0.4 : 1.0;

    return connectDragSource(
      <div
        className='element-draggable'
        style={{ opacity }}
      >
        {children}
      </div>
    );
  }
}

export default DragSource(
  Draggable.type,
  Draggable.spec,
  Draggable.collect,
)(Draggable);
