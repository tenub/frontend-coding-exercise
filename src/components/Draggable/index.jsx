import { useDrag } from 'react-dnd';
import { ItemTypes } from '../../ItemTypes';

const Draggable = ({
  children,
  deleteItem,
  position,
  name,
  type,
}) => {
  const [{ isDragging }, drag] = useDrag(() => ({
    type: ItemTypes[type.toUpperCase()],
    item: { children, position, name, type },
    end: (item, monitor) => {
      if (!deleteItem) {
        return;
      }

      const dropResult = monitor.getDropResult();
      if (item && !dropResult) {
        deleteItem(item.position);
      }
    },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
      handlerId: monitor.getHandlerId(),
    }),
  }))

  const opacity = isDragging ? 0.4 : 1;

  return (
    <div
      ref={drag}
      className='element-draggable'
      style={{ opacity }}
    >
      {children}
    </div>
  );
};

export default Draggable;
