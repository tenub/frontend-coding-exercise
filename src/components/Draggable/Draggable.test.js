import React from 'react';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { DndProvider } from 'react-dnd';
import TestRenderer from 'react-test-renderer';

import Draggable from './';

test('Draggable renders successfully', () => {
  const draggable = (
    <DndProvider backend={HTML5Backend}>
      <Draggable
        index={0}
        name='text'
        type='text'
      >
        Text
      </Draggable>
    </DndProvider>
  );
  const testRenderer = TestRenderer.create(draggable);
  const testTree = testRenderer.toJSON();
  expect(testTree).toMatchSnapshot();
});
