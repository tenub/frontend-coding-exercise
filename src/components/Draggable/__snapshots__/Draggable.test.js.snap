// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable renders successfully 1`] = `
<div
  className="element-draggable"
  style={
    Object {
      "opacity": 1,
    }
  }
>
  Text
</div>
`;
